File: kafka/golang/consumers/kafka_consumer.go
package consumers

import (
	"context"
	"log"

	"github.com/confluentinc/confluent-kafka-go/v2/kafka"
)

type KafkaConsumer interface {
	Close()
}

type kafkaConsumer struct {
	Consumer *kafka.Consumer
}

func NewKafkaConsumer(ctx context.Context) KafkaConsumer {
	config := &kafka.ConfigMap{
		"bootstrap.servers": "localhost:9092",
		"group.id":          "text-message-consumer-group-table",
		"auto.offset.reset": "earliest",
	}
	consumer, err := kafka.NewConsumer(config)
	if err != nil {
		log.Fatalln("error when creating new consumer text-message:", err)
	}

	err = consumer.Subscribe("text-message", nil)
	if err != nil {
		log.Fatalln("error when subscribing consumer:", err)
	}

	go func() {
		for {
			select {
			case <-ctx.Done():
				log.Println("stopping consumer")
				return
			default:
				message, err := consumer.ReadMessage(-1)
				if err != nil {
					log.Println("error reading message:", err)
				} else {
					log.Println("message:", string(message.Value), message.TopicPartition.Partition, message.TopicPartition.Offset)
				}
			}
		}
	}()

	log.Println("listening consumer text-message")
	return &kafkaConsumer{
		Consumer: consumer,
	}
}

func (consumer *kafkaConsumer) Close() {
	err := consumer.Consumer.Close()
	if err != nil {
		log.Fatalln("error when closing kafka consumer:", err)
	}
	log.Println("stop consumer")
}

File: kafka/rust/src/consumers/kafka_consumer.rs
use std::process;

use chrono::Local;
use rdkafka::{consumer::{Consumer, StreamConsumer}, message::Message, ClientConfig};
use tokio_stream::StreamExt;

pub struct KafkaConsumerImpl {

}

impl KafkaConsumerImpl {
    pub fn new() -> KafkaConsumerImpl {
        println!("{} kafka: creating consumer", Local::now().format("%Y-%m-%d %H:%M:%S%.3f").to_string());
        const BROKER: &str = "localhost:9092";
        const TOPIC: &str = "text-messages";
        const GROUP_ID: &str = "text-message-consumer-group-table";
        let consumer: StreamConsumer = match ClientConfig::new()
            .set("bootstrap.servers", BROKER)
            .set("group.id", GROUP_ID)
            .set("auto.offset.reset", "earliest")
            .create() {
                Ok(consumer) => consumer,
                Err(err) => {
                    println!("error creating consumer: {}", err);
                    process::exit(1);
                }
            };
        tokio::spawn(async move {
            consumer.subscribe(&[TOPIC]).expect("cannot subscribe to: topic");
            // consumer need to be static if i put it above tokio::spawn, so just put it in tokio::spawn
            let mut stream = consumer.stream();
            while let Some(Ok(messsage)) = stream.next().await {
                if let Some(payload) = messsage.payload_view::<str>() {
                    println!("Received: {:?} with key {:?}", payload, messsage.key());
                }
            }
        });
        println!("{} kafka: created consumer", Local::now().format("%Y-%m-%d %H:%M:%S%.3f").to_string());
        KafkaConsumerImpl {  }
    }
}